### PR TITLE
Split Typha metrics by syncer type

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

This PR splits Typha's Prometheus metrics by syncer type. Many of Typha's statistics are only meaningful within the scope of a syncer but previously, all the syncers wrote to the same metrics resulting in metrics that would bounce around between the values associated with the different syncers.

## Todos

- [x] Tests
- [x] Documentation
- [x] Release note

## Release Note

```release-note
Many of Typha's Prometheus metrics are now split by syncer (client) type, represented by a label "syncer" on the metrics. This prevents cross-talk where the syncers would all share the same metrics and the last writer to the metric would "win".
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: typha/pkg/syncserver/sync_server.go
Comment (caseydavenport) [inline typha/pkg/syncserver/sync_server.go:558]: Do we need a nil check here as well? I see the Dec() call above is protected.

Comment (caseydavenport) [reply typha/pkg/syncserver/sync_server.go:558]: I suppose moving this down below the handshake changes this gauge's meaning as well, but that's probably OK.

**Review by caseydavenport — APPROVED** [review]

File: typha/pkg/syncserver/sync_server.go
Comment (fasaxc) [reply typha/pkg/syncserver/sync_server.go:558]: No, the gauge is set up by `doHandshake()` so no nil check needed but the defer block runs even if we don't get this far.  Suppose I could do `defer h.gaugeNumConnections.Dec()` here instead but previously we only decremented after all connection resources were freed.

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Comment by marvin-tigera** [comment]
Removing "merge-when-ready" label due to new commits

File: typha/pkg/syncserver/sync_server.go
Comment (fasaxc) [reply typha/pkg/syncserver/sync_server.go]: Hmm, some tests were broken by that change (and they were using it to test the interaction with handshake).  I'll reinstate the old stat and add a new one for each syncer type.

**Comment by fasaxc** [comment]
Failures look to be from tests that don't use typha in any way so writing off as flakes.

**Pull request merged**
